meta/material.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return err
+	}
+	defer stm.Close()
+	rs, err := stm.Exec(material.Description, typeID, aliasID)
+	if err != nil {
+		return err
+	}
+	//update tagsofmaterial,delete existing records, add new ones
+	//delete
+	stm, err := tx.Prepare("delete from tagsofmaterial where aliasid=?")
+	if err != nil {
+		return err
+	}
+	defer stm.Close()
+	rs, err := stm.Exec(aliasID)
+	if err != nil {
+		return err
+	}
+	//add
+	if material.Tags != nil && len(material.Tags) != 0 {
+		allTag := strings.Split(material.Tags, ";")
+		for i := 0; i < len(allTag); i++ {
+			//get tagid
+			row := db.QueryRow("select id from tag where name=?", allTag[i])
+			var tagID int64 = -1
+			err := row.Scan(&tagID)
+			if err != nil {
+				return err
+			}
+			//insert into tagsofmaterial
+			stm, err := tx.Prepare("insert into tagsofmaterial (aliasid, tagid) values (?, ?)")
+			if err != nil {
+				return err
+			}
+			defer stm.Close()
+			rs, err := stm.Exec(aliasID, tagID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func deleteMaterial(db *sql.DB, materialName string) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+	//get aliasid
+	row := db.QueryRow("select id from alias where materialname=?", materialName)
+	var aliasID int64 = -1
+	err := row.Scan(&aliasID)
+	if err != nil {
+		return err
+	}
+	//delete material
+	stm, err := tx.Prepare("delete from material where aliasid=?")
+	if err != nil {
+		return err
+	}
+	defer stm.Close()
+	rs, err := stm.Exec(aliasID)
+	if err != nil {
+		return err
+	}
+	//delete tagofmaterial
+	stm, err := tx.Prepare("delete from tagsofmaterial where aliasid=?")
+	if err != nil {
+		return err
+	}
+	defer stm.Close()
+	rs, err := stm.Exec(aliasID)
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getMaterial(db *sql.DB, materialName string) (*Material, error) {
+	tx, err := db.Begin()
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+	//get  type & description
+	row := db.QueryRow("select material.description, materialtype.typename  from material, materialtype, alias where alias.materialname=? and alias.id=material.aliasid and material.typeid=materialtype.id", materialName)
+	var description string
+	var typename string
+	err := row.Scan(&description, &typename)
+	if err != nil {
+		return nil, err
+	}
+	//get all alias
+	rows, err := store.DB.Query("select materialname from alias where id=(select id from alias where materialname=?)", materialName)
+	defer rows.Close()
+	var allalias string = ""
+	for rows.Next() {
+		var aliastmp string
+		err = rows.Scan(&aliastmp)
+		if err != nil {
+			return nil, err
+		}
+		if strings.Compare(materialName, aliastmp) != 0 {
+			if allalias == "" {
+				allalias = aliastmp
+			} else {
+				allalias += ";" + aliastmp
```

Comment:
Could we use some const variables instead of hardcoding the magic strings?

Reply:
I just added dataSourceName in a JSON file.